### PR TITLE
Ibp 3950 inventory manager merge lots

### PR DESCRIPTION
**Pull request by lik-83**

### CODE REVIEW

File: src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java [hunk 1]
```diff
@@ -37,14 +26,7 @@
...
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Date;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java:29]: Please review your IDE import comfiguration @lik-83

File: src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java [hunk 2]
```diff
@@ -127,6 +122,73 @@ public void lotWithNoOpenBalanceClosed_Ok() {
 
 	}
 
+	@Test
+	public void shouldMergeLots() {
```
Comment (clarysabel) [inline src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java:126]: to be consistent, I will call it testMergeLots_Ok @lik-83

File: src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java [hunk 3]
```diff
@@ -14,25 +14,20 @@
 import org.generationcp.middleware.pojos.Germplasm;
 import org.generationcp.middleware.pojos.LocationType;
 import org.generationcp.middleware.pojos.UDTableType;
-import org.generationcp.middleware.pojos.ims.EntityType;
-import org.generationcp.middleware.pojos.ims.Lot;
-import org.generationcp.middleware.pojos.ims.LotStatus;
-import org.generationcp.middleware.pojos.ims.Transaction;
-import org.generationcp.middleware.pojos.ims.TransactionStatus;
-import org.generationcp.middleware.pojos.ims.TransactionType;
+import org.generationcp.middleware.pojos.ims.*;
```
Comment (clarysabel) [inline src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java:17]: @lik-83 same here with the imports

File: src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java [hunk 4]
```diff
@@ -164,4 +226,14 @@ private void resolveStorageLocation() {
 		this.storageLocationId = this.daoFactory.getLocationDAO().getDefaultLocationByType(id).getLocid();
 	}
 
+	private void assertTransaction(Transaction actualTransaction, TransactionType type, TransactionStatus status,
```
Comment (clarysabel) [inline src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java:229]: to be consistent I will add final to the params @lik-83

File: src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java [hunk 5]
```diff
@@ -248,6 +230,28 @@ public void closeLots(final Integer userId, final List<Integer> lotIds) {
 		this.daoFactory.getLotDao().closeLots(lotIds);
 	}
 
+	@Override
+	public void mergeLots(final Integer userId, final Integer keepLotId, final LotsSearchDto lotsSearchDto) {
+		//Search selected lots to be merged and remove the one to keep
+		List<ExtendedLotDto> extendedLotDtos = this.searchLots(lotsSearchDto, null).stream()
```
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java:236]: we also use final for the variables, something we need to discuss but it is the standard we have until now

File: src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java (hunk 1 above)
Comment (lik-83) [reply src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java]: Done!

File: src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java (hunk 3 above)
Comment (lik-83) [reply src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java]: Done!

File: src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java (hunk 4 above)
Comment (lik-83) [reply src/test/java/org/generationcp/middleware/service/impl/inventory/LotServiceImplIntegrationTest.java]: Done!

File: src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java [hunk 6]
```diff
@@ -248,6 +249,28 @@ public void closeLots(final Integer userId, final List<Integer> lotIds) {
...
+	@Override
+	public void mergeLots(final Integer userId, final Integer keepLotId, final LotsSearchDto lotsSearchDto) {
+		//Search selected lots to be merged and remove the one to keep
+		final List<ExtendedLotDto> extendedLotDtos = this.searchLots(lotsSearchDto, null).stream()
+				.filter(extendedLotDto -> extendedLotDto.getLotId().intValue() != keepLotId.intValue())
+				.collect(Collectors.toList());
+		//Create a transaction for each of the discarded lots
+		extendedLotDtos.stream()
+			.forEach(extendedLotDto -> {
+				final Transaction transaction = new Transaction(TransactionType.DEPOSIT, TransactionStatus.CONFIRMED,
```
[the remark this answers is not in the document]
Comment (lik-83) [reply src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java:261]: Ok!

[the remark this answers is not in the document]
Comment (lik-83) [reply src/main/java/org/generationcp/middleware/service/impl/inventory/LotServiceImpl.java:261]: Done!

**Pull request merged**